lib/worker-config.js

```diff
@@ -0,0 +1,293 @@
+    if (typeof id !== 'string') {
+      this._throw(errors.InvalidWorkerConfiguration, 'id must be provided');
+    }
+    this.id = id;
+  }
+ 
+  /**
+   * Standardize exceptions thrown
+   */
+  _throw(code, msg) {
+    throw new code(msg, {
+      'class': this.constructor.name, 
+      id: this.id,
+    });
+  }
+
+  /**
+   * Evaluate a worker configuration
+   */
+  evaluate(satisfiers) {
+    this._throw(errors.MethodUnimplemented, 'WorkerConfiguration.evaluate()');
+  }
+
+  /**
+   * Must return an iterator of strings
+   */
+  workerTypes() {
+    this._throw(errors.MethodUnimplemented, 'WorkerConfiguration.workerTypes()');
+  }
+}
+
+/**
+ * This class represents a static worker configuration in memory.  This class 
+ * does not support any automatic provisioning, instead having a constant
+ * configuration.  The configuration must be equal to the result of evaluating
+ * a successful WorkerConfiguration.evaluate().
+ */
+class StaticWorkerConfiguration extends WorkerConfiguration {
+  constructor({id, workerTypes, configuration}) {
+    super({id});
+
+    if (!Array.isArray(workerTypes)) {
+      this._throw(errors.InvalidWorkerConfiguration, 'workerTypes must be array');
+    }
+
+    for (let workerType of workerTypes) {
+      if (typeof workerType !== 'string') {
+        this._throw(errors.InvalidWorkerConfiguration, 'workerTypes entries must be strings');
+      }
+    }
+
+    this._workerTypes = workerTypes;
+    if (typeof configuration !== 'object') {
+        this._throw(errors.InvalidWorkerConfiguration, 'configuration must be object');
+    }
+    this.configuration = configuration;
+  }
+
+  /**
+   * Evaluate a worker configuration
+   */
+  evaluate(satisfiers) {
+    return this.configuration;
+  }
+
+  workerTypes() {
+    return this._workerTypes.values();
+  }
+}
+
+/**
+ * This class represents a provisioned worker configuration in memory. *
+ *
+ * The constructor parameters are:
+ *  - id: identifier for this worker configuration
+ *  - workerTypes: list of objects {workerType, biddingStrategy, providers}
+ *  - rules: A Ruleset to use with this worker type
+ *  - biddingStrategies: reference to a Map which maps bidding strategy
+ *    identifier to the bidding strategy implmenentation
+ *  - providers: reference to a Map which maps provider identifier to the
+ *    provider implementation
+ */
+class ProvisionedWorkerConfiguration extends WorkerConfiguration {
+  constructor({id, workerTypeConfigurations, rules}) {
+    super({id});
+
+    for (let workerTypeConfiguration of workerTypeConfigurations) {
+      let workerType = workerTypeConfiguration.get('workerType');
+      let providerIds = workerTypeConfiguration.get('providerIds');
+      let biddingStrategyId = workerTypeConfiguration.get('biddingStrategyId');
+
+      if (typeof workerType !== 'string') {
+        this._throw(errors.InvalidWorkerConfiguration, 'worker type name must be string');
+      }
+      this.workerType = workerType;
```

Comment:
hmm, it was always a for-of, but I think it's a typo.  I've removed the assignments to `this` in this loop in my current patch set.

Reply:
Ahh, I remember, this code was originally in a `.map` in the factory function, where assignment was important, but I decided to move the validation into the constructor so that both factory-built instances had the same validation as those built with the constructor.